index.js

```diff
@@ -13,140 +13,163 @@ const defaultOptions = {
-    // occurred in this handler, other plugins may deny the request since the browser will
-    // remove most headers (such as the Authentication header).
-    //
-    // This route simply enables fastify to accept preflight requests.
-    fastify.options('*', { schema: { hide: hideOptionsRoute } }, (req, reply) => {
-      if (!req.corsPreflightEnabled) {
-        // Do not handle preflight requests if the origin option disabled CORS
-        reply.callNotFound()
-        return
-      }
 
-      reply.send()
+  let hideOptionsRoute = true
+  if (typeof opts === 'function') {
+    handleCorsOptionsDelegator(opts, fastify)
+  } else {
+    hideOptionsRoute = opts.hideOptionsRoute
+    const corsOptions = Object.assign({}, defaultOptions, opts)
+    fastify.addHook('onRequest', (req, reply, next) => {
+      onRequest(fastify, corsOptions, req, reply, next)
     })
   }
 
-  next()
+  // The preflight reply must occur in the hook. This allows fastify-cors to reply to
+  // preflight requests BEFORE possible authentication plugins. If the preflight reply
+  // occurred in this handler, other plugins may deny the request since the browser will
+  // remove most headers (such as the Authentication header).
+  //
+  // This route simply enables fastify to accept preflight requests.
+  fastify.options('*', { schema: { hide: hideOptionsRoute } }, (req, reply) => {
+    if (!req.corsPreflightEnabled) {
+      // Do not handle preflight requests if the origin option disabled CORS
+      reply.callNotFound()
+      return
+    }
 
-  function onRequest (req, reply, next) {
-    // Always set Vary header
-    // https://github.com/rs/cors/issues/10
-    vary(reply, 'Origin')
+    reply.send()
+  })
 
-    resolveOriginOption(req, (error, resolvedOriginOption) => {
-      if (error !== null) {
-        return next(error)
-      }
+  next()
+}
 
-      // Disable CORS and preflight if false
-      if (resolvedOriginOption === false) {
-        return next()
+function handleCorsOptionsDelegator (optionsResolver, fastify) {
+  fastify.addHook('onRequest', (req, reply, next) => {
+    if (optionsResolver.length === 2) {
+      handleCorsOptionsCallbackDelegator(optionsResolver, fastify, req, reply, next)
+      return
+    } else {
+      // handle delegator based on Promise
+      const ret = optionsResolver(reply)
+      if (ret && typeof ret.then === 'function') {
+        ret.then(options => Object.assign({}, defaultOptions, options))
+          .then(corsOptions => onRequest(fastify, corsOptions, req, reply, next)).catch(next)
+        return
       }
+    }
+    next(new Error('Invalid CORS origin option'))
+  })
+}
 
-      // Falsy values are invalid
-      if (!resolvedOriginOption) {
-        return next(new Error('Invalid CORS origin option'))
-      }
+function handleCorsOptionsCallbackDelegator (optionsResolver, fastify, req, reply, next) {
+  optionsResolver(req, (err, options) => {
```

Comment:
Please handle the case if optionsHandler is an async function/returns a promise.

Reply:
Check line 58 @mcollina